scalalib/src/mill/scalalib/TestModuleUtil.scala

```diff
@@ -173,15 +209,217 @@ private[scalalib] object TestModuleUtil {
+
+    def prepareTestSelectorFolder(selectors2: Seq[String], base: os.Path): os.Path = {
+      val selectorFolder = base / "selectors"
+      os.makeDir.all(selectorFolder)
+      selectors2.zipWithIndex.foreach { case (s, i) =>
+        os.write.over(selectorFolder / s, Array.empty[Byte])
+      }
+      selectorFolder
+    }
+
+    def runTestRunnerSubprocess(
+        groupIndex: Int,
+        processIndex: Int,
+        base: os.Path,
+        selectorFolder: os.Path,
+        force: Boolean,
+        logger: Logger
+    ) = {
+      if (force || os.list(selectorFolder).nonEmpty) {
+        val stealFolder = base / "steal"
+        os.makeDir.all(stealFolder)
+        val stealLog = stealFolder / os.up / "steal.log"
+        os.write.over(stealLog, Array.empty[Byte])
+        stealLogMap.put(
+          (groupIndex, processIndex),
+          (stealLog, logger.ticker)
+        )
+        val result = callTestRunnerSubprocess(base, Right(selectorFolder, stealFolder))
+        stealLogMap.remove((groupIndex, processIndex))
+        Some(result)
+      } else {
+        None
+      }
+    }
+
+    val groupFolderData = filteredClassLists match {
+      case Nil => Seq((Task.dest, prepareTestSelectorFolder(Nil, Task.dest), 0, ""))
+      case Seq(singleTestClassList) =>
+        Seq((
+          Task.dest,
+          prepareTestSelectorFolder(singleTestClassList, Task.dest),
+          singleTestClassList.length,
+          ""
+        ))
+      case multipleTestClassLists =>
+        val maxLength = multipleTestClassLists.length.toString.length
+        multipleTestClassLists.zipWithIndex.map { case (testClassList, i) =>
+          val groupPromptMessage = testClassList match {
+            case Seq(single) => single
+            case multiple =>
+              TestModuleUtil.collapseTestClassNames(
+                multiple
+              ).mkString(", ") + s", ${multiple.length} suites"
+          }
+          val paddedIndex = mill.internal.Util.leftPad(i.toString, maxLength, '0')
+          val folderName = testClassList match {
+            case Seq(single) => single
+            case multiple =>
+              s"group-$paddedIndex-${multiple.head}"
+          }
+
+          (
+            Task.dest / folderName,
+            prepareTestSelectorFolder(testClassList, Task.dest / folderName),
+            testClassList.length,
+            groupPromptMessage
+          )
+        }
+    }
+
+    val jobs = Task.ctx() match {
+      case j: Ctx.Jobs => j.jobs
+      case _ => 1
+    }
+
+    val maxProcessLength = jobs.toString.length
+    val groupLength = groupFolderData.length
+    val maxGroupLength = groupLength.toString.length
+
+    val subprocessFutures = for {
+      processIndex <- 0 until jobs
+      groupIndex <- 0 until groupLength
+      (baseFolder, selectorFolder, numTests, groupPromptMessage) = groupFolderData(groupIndex)
+      if (processIndex == 0) || (processIndex < numTests)
```

Comment:
Rather than doing this filter, it would be simpler to instead loop over `processIndex <- 0 until math.min(math.max(jobs, 1) numTests)` to just pick the correct number of processes up front